File: 프로그래머스/1/42889. 실패율/실패율.py
def solution(N, stages):
    answer = []
    clear = [0] * (N + 2)
    counts = [0] * (N + 2)
    failRateList = []
    
    for i in range(len(stages)):
        stage = stages[i]
        counts[stage] += 1
        for j in range(1, stage+1):
            clear[j] += 1
    
    for i in range(1, N+1):
        if counts[i] == 0:
            failRateList.append([i, 0])
        else:
            failRateList.append([i, counts[i] / clear[i]])
    
    failRateList.sort(key=lambda x : -x[1])
    print(failRateList)
    
    for i, data in failRateList:
        answer.append(i)
    
    return answer

File: 프로그래머스/1/42889. 실패율/실패율.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class Solution {
    public int[] solution(int N, int[] stages) {
        int[] answer = new int[N];
		
		int[] ratio = new int[N+1];
		List<Double> loseRatio = new ArrayList<>();
		double totalPlayer = stages.length;
		
		for(int stage : stages) {
			ratio[stage-1]++;
		}
		
		for(int i=0; i<answer.length; i++) {
			if(totalPlayer != 0) {
				loseRatio.add(ratio[i] / totalPlayer);
				totalPlayer -= ratio[i];
			} else {
				loseRatio.add(0.0);
			}
		}
		
		List<Double> temp = new ArrayList<>(loseRatio);
		
		temp.sort(Collections.reverseOrder());

		int n = 0;
		for(Double d : temp) {
			int index = loseRatio.indexOf(d);
			answer[n] = index+1;
			n++;
			loseRatio.set(index, -1.0);
		}
        
        return answer;
    }
}